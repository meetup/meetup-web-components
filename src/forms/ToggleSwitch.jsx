import PropTypes from 'prop-types';
import React from 'react';

import { Toggle } from '@meetup/swarm-components';

/**
 * @module ToggleSwitch
 */
class ToggleSwitch extends React.Component {
	render() {
		const { isActive, label, labelledBy, id, name, ...other } = this.props;

		const labelProps = labelledBy ? { 'aria-labelledby': labelledBy } : {};

		return (
			<div>
				{label && <label htmlFor={id || name}>{label}</label>}
				<Toggle
					id={id || name}
					name={name}
					checked={isActive}
					{...labelProps}
					{...other}
				/>
			</div>
		);
	}
}

ToggleSwitch.propTypes = {
	/** Whether to use disabled attribute on the button element and apply disabled switch styles */
	disabled: PropTypes.bool,

	/** Whether the switch is toggled on */
	isActive: PropTypes.bool,

	/** What we render into the input's `<label />` */
	label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),

	/** The id of the element that labels the toggle switch */
	labelledBy: PropTypes.string,

	/** The identifier used to associate a label with a toggle switch */
	name: PropTypes.string.isRequired,
};

export default ToggleSwitch;
