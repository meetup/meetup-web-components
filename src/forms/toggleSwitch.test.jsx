import React from 'react';
import { shallow, mount } from 'enzyme';

import ToggleSwitch from './ToggleSwitch';

describe('ToggleSwitch', function() {
	const DEFAULT_PROPS = {
		name: 'testToggle',
	};

	const shallowWrapper = (props = {}) =>
		shallow(<ToggleSwitch {...DEFAULT_PROPS} {...props} />);
	const mountWrapper = (props = {}) =>
		mount(<ToggleSwitch {...DEFAULT_PROPS} {...props} />);

	describe('basic', function() {
		it('renders inactive toggle switch', function() {
			expect(shallowWrapper()).toMatchSnapshot();
		});

		it('renders checked toggle switch', function() {
			expect(shallowWrapper({ checked: true })).toMatchSnapshot();
		});

		it('has a label only if one is given', () => {
			const component = mountWrapper({ label: 'Turn it on?' });
			expect(component.find('label')).not.toBe(null);
		});

		it('sets `aria-labelledby` attribute if one is given', () => {
			const CUSTOM_LABELLEDBY = 'test-label';
			const component = mountWrapper({ labelledBy: CUSTOM_LABELLEDBY });

			expect(component.find(`button`).prop('aria-labelledby')).toBe(
				CUSTOM_LABELLEDBY
			);
		});
	});

	describe('disabled', function() {
		it('sets aria-readonly attribute', function() {
			const component = mountWrapper({ disabled: true });
			expect(component.find(`button`).props()['aria-readonly']).toBe(true);
		});

		it('does not update aria-checked when clicked when disabled', function() {
			const component = mountWrapper({ disabled: true });
			const btnNode = component.find(`button`);

			expect(btnNode.props()['aria-checked']).not.toBe(true);
			btnNode.simulate('click');
			expect(btnNode.props()['aria-checked']).not.toBe(true);
		});
	});
});
