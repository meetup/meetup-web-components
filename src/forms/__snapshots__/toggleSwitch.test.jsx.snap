// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitch basic renders checked toggle switch 1`] = `
<div>
  <Toggle
    checked={true}
    id="testToggle"
    name="testToggle"
  />
</div>
`;

exports[`ToggleSwitch basic renders inactive toggle switch 1`] = `
<div>
  <Toggle
    id="testToggle"
    name="testToggle"
  />
</div>
`;
